yozhao101/sonic_buildimage
# [Service] Restart Teamd service upon unexpected critical process exited.

**- What I did**
Restart teamd service and also its dependent service which is docker-dhcp-relay service if one of
critical processes running in teamd container exited abnormally.  



**- How I did it**
Generally I follow the framework created by Joe to implement this feature in Teamd container.

1. Add supervisor-proc-exit-listener event listener plugin for Supervisor in Teamd Docker 
container which in turn loads a list of critical processes for which to monitor for unexpected 
exits.

2. Configure temad.service to always auto-restart the service if it stops, with a delay of 30 seconds 
Also set a rate limit of 3 restarts within 20 minutes (1200 seconds).

**- How to verify it**
On your switch device, please use _docker ps_ command to list the running docker containers.
Then use _docker exec -it container_id /bin/bash_ to login your container. Type _top_ command
on the shell will display all the processes dynamically and you will spot the process id of one
of the critical processes. Finally type the command _kill -9 process_id_ to kill one process.

After exiting the container, you can use _watch -n 1 docker ps_ to dynamically see the restart
of teamd and dhcp-relay containers.

**- What I did not**
Actually there are three critical processes in teamd container: **teamd**, **teammgrd** and **teamsyncd**. 
Since teammgrd and teamsyncd are monitored by supervisord process, if one of these two 
processes exited accidentally, the supervisord will be killed by SIGTERM sent by exit event listener
and then teamd container will be restarted.

However, teamd process is spawned by teammgrd process and it is not monitored by 
supervisord. As such, if teamd process exited, teamd container will not be automatically
restarted. This is a issue and currently is being discussed.
